tests/compat/test_requests.py

```diff
@@ -0,0 +1,2285 @@
+        """Tests resolve_redirect doesn't fail when merging cookies
+        with non-RequestsCookieJar cookiejar.
+
+        See GH #3579
+        """
+        cj = cookiejar_from_dict({"foo": "bar"}, cookielib.CookieJar())
+        s = requests.Session()
+        s.cookies = cookiejar_from_dict({"cookie": "tasty"})
+
+        # Prepare request without using Session
+        req = requests.Request("GET", httpbin("headers"), cookies=cj)
+        prep_req = req.prepare()
+
+        # Send request and simulate redirect
+        resp = s.send(prep_req)
+        resp.status_code = 302
+        resp.headers["location"] = httpbin("get")
+        redirects = s.resolve_redirects(resp, prep_req)
+        resp = next(redirects)
+
+        # Verify CookieJar isn't being converted to RequestsCookieJar
+        assert isinstance(prep_req._cookies, cookielib.CookieJar)
+        assert isinstance(resp.request._cookies, cookielib.CookieJar)
+        assert not isinstance(resp.request._cookies, requests.cookies.RequestsCookieJar)
+
+        cookies = {}
+        for c in resp.request._cookies:
+            cookies[c.name] = c.value
+        assert cookies["foo"] == "bar"
+        assert cookies["cookie"] == "tasty"
+
+    def test_requests_in_history_are_not_overridden(self, httpbin):
+        resp = requests.get(httpbin("redirect/3"))
+        urls = [r.url for r in resp.history]
+        req_urls = [r.request.url for r in resp.history]
+        assert urls == req_urls
+
+    def test_history_is_always_a_list(self, httpbin):
+        """Show that even with redirects, Response.history is always a list."""
+        resp = requests.get(httpbin("get"))
+        assert isinstance(resp.history, list)
+        resp = requests.get(httpbin("redirect/1"))
+        assert isinstance(resp.history, list)
+        assert not isinstance(resp.history, tuple)
+
+    @pytest.mark.httpx_bug
+    @pytest.mark.xfail(reason="'NoneType' object has no attribute 'encode'")
+    def test_headers_on_session_with_None_are_not_sent(self, httpbin):
+        """Do not send headers in Session.headers with None values."""
+        ses = requests.Session()
+        ses.headers["Accept-Encoding"] = None
+        resp = ses.get(httpbin("get"))
+        assert "Accept-Encoding" not in resp.request.headers
+
+    def test_headers_preserve_order(self, httpbin):
+        """Preserve order when headers provided as OrderedDict."""
+        ses = requests.Session()
+        ses.headers = collections.OrderedDict()
+        ses.headers["Accept-Encoding"] = "identity"
+        ses.headers["First"] = "1"
+        ses.headers["Second"] = "2"
+        headers = collections.OrderedDict([("Third", "3"), ("Fourth", "4")])
+        headers["Fifth"] = "5"
+        headers["Second"] = "222"
+        resp = ses.get(httpbin("get"), headers=headers)
+        items = list(resp.request.headers.items())
+
+        # Remove headers automatically added by HTTPX.
+        while items[0][0] != "accept-encoding":
```

Comment:
Not sure — *if* the test case should pass (e.g. order is preserved), then all headers in the list before the first one we added (`Accept-Encoding`) can be removed, so that we're only left with the ones we want to check the order of. Besides I don't think we want to couple this with other HTTPX's behavior.

Reply:
You're right, let's leave as-is.